### PR TITLE
Add the ability to append and prepend middleware priority from the application builder

### PR DESCRIPTION
**Pull request by ollieread**
This is a follow-up PR for #52897 that adds two methods to the application builder, allowing for easy access to the add middleware after/before methods on the kernel.

Using the examples from the previous PR, the methods would be accessible like so:

```php
return Application::configure(basePath: dirname(__DIR__))
    ->withRouting(
        web: __DIR__.'/../routes/web.php',
        commands: __DIR__.'/../routes/console.php',
        health: '/up',
    )
    ->withMiddleware(function (Middleware $middleware) {
        $middleware->appendToPriorityList(
		    [
		        \Illuminate\Cookie\Middleware\EncryptCookies::class,
		        \Illuminate\Contracts\Auth\Middleware\AuthenticatesRequests::class,
		    ],
		    \Illuminate\Routing\Middleware\ValidateSignature::class
		);

		$middleware->prependToPriorityList(
		    [
		        \Illuminate\Cookie\Middleware\EncryptCookies::class,
		        \Illuminate\Contracts\Auth\Middleware\AuthenticatesRequests::class,
		    ],
		    \Illuminate\Routing\Middleware\ValidateSignature::class
		);
    })
    ->withExceptions(function (Exceptions $exceptions) {
        //
    })->create();
```

They map as follows:

 - `Middleware::appendToPriorityList()` -> `Kernel::addToMiddlewarePriorityAfter()`
 - `Middleware::prependToPriorityList()` -> `Kernel::addToMiddlewarePriorityBefore()`

### CODE REVIEW

**Comment by taylorotwell** [comment]
Can the signature be something like:

```php
$middleware->appendToPriorityList(after: $middlewareToPutTheNewMiddlewareAfter, append: $theMiddlewareToAppend);
$middleware->prependToPriorityList(before: $middlewareToPutTheNewMiddlewareBefore, prepend: $theMiddlewareToPrepend);
```

Please mark as ready for review when the requested changes have been made.

**Comment by taylorotwell** [comment]
Hey there! I'm not sure if the parameter names match what I mentioned?

**Comment by ollieread** [comment]
> Hey there! I'm not sure if the parameter names match what I mentioned?

Oh sorry, I had assumed they were just to illustrate as the names are quite long. No worries, I'll update that now.

**Comment by ollieread** [comment]
> Hey there! I'm not sure if the parameter names match what I mentioned?

All done!

**Comment by taylorotwell** [comment]
Sorry, I mean the actual parameter names, which are here:

CleanShot 2024-10-30 at 09 57 11@2x

**Comment by ollieread** [comment]
> Sorry, I mean the actual parameter names, which are here:
> 
> CleanShot 2024-10-30 at 09 57 11@2x

Ah, that makes so much more sense. Sorry, that's on me, my head is all over the place today!

File: src/Illuminate/Foundation/Configuration/ApplicationBuilder.php [hunk 1]
```diff
@@ -269,6 +269,18 @@ public function withMiddleware(?callable $callback = null)
...
+                foreach ($priorityAppends as $middleware => $after) {
+                    $kernel->addToMiddlewarePriorityAfter($after, $middleware);
+                }
+            }
+
+            if ($priorityPrepends = $middleware->getMiddlewarePriorityPrepends()) {
+                foreach ($priorityPrepends as $middleware => $before) {
+                    $kernel->addToMiddlewarePriorityBefore($before, $middleware);
+                }
+            }
```
Comment (patrickomeara) [inline src/Illuminate/Foundation/Configuration/ApplicationBuilder.php:283]: When using both `prependToPriorityList()` and `appendToPriorityList()` the `$middleware` variable is overridden on line 274. `getMiddlewarePriorityPrepends()` is then called on a string.

**Review by patrickomeara — COMMENTED** [review]
I think this needs tests as I have tried to move my `->priority()` call to use this but the priority doesn't change.

Additionally there is an issue with using both as below.

File: src/Illuminate/Foundation/Configuration/ApplicationBuilder.php (hunk 1 above)
Comment (ollieread) [reply src/Illuminate/Foundation/Configuration/ApplicationBuilder.php:283]: Ah this seems to be a bug either prepending!

**Comment by ionutantohi** [comment]
Do you think this can be extended to accept an array of middlewares for append/prepend?

```php
$middleware->appendToPriorityList(
    after: SubstituteBindings::class,
    append: [
        MiddlewareA::class,
        MiddlewareB::class,
    ],
);
```

This reads as: append MiddlewareA and MiddlewareB in that order after SubstituteBindings.

**Pull request merged**